Add quiet-foralls proposal

simonpj opened the pull request:
Propose quiet foralls.

simonpj commented:
@int-index we'd love to have your views here.

AntC2 commented:
> This proposal was developed during a three-day walk in the Italian Alps, at 2500m.

So you were 'high'. Does that explain all the 'high-oglyphics'? (Congratulations on getting out amongst the mountains.)

> * Infrerred, written `forall {{k}. type`, exactly as Inferred forall-binders today.
> * Quiet, written `forall [k}. type`, is new in this proposal.

I hope those seemingly unmatched brackets are typos. But TBH this syntax is now so weird, it no longer looks like Haskell; I have no intuition what would be right, or what it would mean.

This is the reason I no longer use GHC.

int-index commented:
> @int-index we'd love to have your views here.

Like @AntC2 I am somewhat startled by unmatched brackets, so I will assume those are typos.

With that out of the way, I have two actual criticisms of this proposal

1. It adds complexity to an already complex specificity system without addressing existing problems like [#22648 "Order of inferred quantification is observable"](https://gitlab.haskell.org/ghc/ghc/-/issues/22648).

   I feel that any attempt to change/extend specificity should start by addressing [#22648](https://gitlab.haskell.org/ghc/ghc/-/issues/22648), as it reveals that the "Inferred" specificity doesn't even serve its original purpose. We need to figure out how to fix specificity before extending it.

3. Asymmetry between binding sites and application sites is a very dubious idea.

   Those new "quiet" binders come in a predictable, user-specified order, and can be bound at definition sites with `@{a}`. So _at definition sites_, they are basically just another variety of specified binders, right? And I imagine we'll need a way to match up a mix of `@a` and `@{a}` binders with foralls in the signature, in case the user writes a mix of them like `f = \ @{a} @b @{c} @d -> ...`. Therefore, there must be a matching-up algorithm for the two varieties of `@`-binders. So what exactly is achieved by disallowing the `@{a}` at use sites? Someone will inevitably have a use case where they want to specify one of those "quiet" variables, they'll want to write a mixed type application `f @{a} @b @{c} @d`. And there's an algorithm that would allow it (because it's used at definition sites). But it's going to be disallowed as a matter of principle? I don't see what's being gained here.

Hmm. I suppose the one thing that I do like about this proposal is that it was written in the Italian Alps 😄 I'm not yet sold on the rest of it.

VitWW commented:
I really don't understand **why** "quiet forall" even needed.
Just allowing forbidden syntax `\ @{k} ->` and `@{k}` covers most cases shown as examples in this Proposal.
```haskell
f2 :: (forall {k}. forall (a::k). Proxy a -> Int) -> Int
f2 (\ @a -> blah)
f2 (\ @{k} -> \ @(a::k) -> blah)

g :: forall {k}. forall (a::k). blah
g @{Type} @Int
g @{_} @Int
```

AntC2 commented:
This seems just a wrong line of argument:

> you may write the call `g @Int`. But you may not write the call `g @Type @Int`.

Well ok, write `g @(Int :: Type)`.

> Why shouldn't you be able to write this? `f2 (\ @{k} -> \ @(a::k) -> blah)`
> [**] [I've elided lines there, and BTW the next line has more unbalanced brackets -- too much Amaro Alpino?]

Then usage: `f2 (\ @(a::k) -> blah)`; or `f2 @(Int :: Type) ...`; or `f2 @(_ :: Type) ...` if you want to omit the type.

What's behind my distaste is: the meaning of `@` seems to be drifting into ad-hockery. When `@` was introduced, Richard was firmly of the view it _doesn't_ mean "here comes a type" but rather "here comes an optionally omittable element"; merely at the time all omittable elements happened to be types. Then

> `h = \ @{k1} -> \ @{k2} -> \ @(a::k1) -> \ @(b::k2) -> blah`

?? The braces `{k1}` say `k1` is inferred not specified so can't be written at a use site; the `@( ... )` overrides that to say it _can_ be written at a use site ... emm, but it's still omittable; the `\ ... -> ...` says (something) must be an argument ???

I kinda feel the omittability annotation should apply to the lambda `@\ ... ->`, not to the content.

[**] Because it's ugly?

> ... would be tiresome ...

People who want to program in line-noise should be forced to use maximally verbose and tiresome line-noise, to encourage them to find a more elegant way unless they really, really insist.

> The proposal is easy to implement.

Pleeease don't go making unquantified claims like that. You'll only encourage copy-cat behaviour.

tomjaguarpaw commented on proposals/0000-quiet-foralls.rst:
```diff
@@ -176,15 +181,56 @@ what ``-XTypeAbstractions`` does.  Old programs witll continue to
+  be much simpler to just have Required foralls and nothing else -- but then all type arguments
+  would be compulsory, and no one would want to use the language.  It's all about using
+  perhaps-ad-hoc mechanisms to make programming convenient.  This proposal just fills out a
+  missing cornder of the design space.
```
```suggestion
  missing corner of the design space.
```

simonpj commented:
Thanks for pointing out the mis-matched brackets, now fixed.

> Just allowing forbidden syntax \ @{k} -> and @{k} covers most cases shown as examples in this Proposal.

No, that would be bad.  We attempted to explain this in the "TypeAbstractions" bullet of Section 1.1.  Perhaps you can help us explain it better?


> It adds complexity to an already complex specificity system without addressing existing problems like #22648 "Order of inferred quantification is observable".

Wow thanks for that pointer.  I have added a sub-section to the discussion section, pointing to that ticket, and discussing it.

> So what exactly is achieved by disallowing the @{a} at use sites?

Ah. I hadn't thought of that.  We could allow curly braces at use sites too.  So if ``f :: forall {k}. forall (a::k). Proxy a -> Int`` we could call it thus:

* ``f Proxy``
* ``f @Int Proxy``
* ``f @{Type} @Int Proxy``
* ``f @{Type} Proxy``

That indeed seems plausible.

maralorn commented on proposals/0000-quiet-foralls.rst:
```diff
@@ -0,0 +1,253 @@
+
+thereby bringing both ``a`` and ``k`` into scope. But
+
+* The says that using a pattern signature is unnsatisfactory here.
```
I assume this sentence is missing a word?

int-index commented:
@simonpj Perhaps there's something I'm missing, but doesn't this proposal just shift all the issues of today's inferred foralls `forall {k}.` to the new inferred foralls `forall {{k}}.`? There wouldn't be `@{{k}}` binders, so all the tickets mentioned in the motivation and all the observed issues could be reproduced by simply adding a second pair of braces. E.g. you link to [MR !13190](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/13190) which has trouble with this example

```haskell
import Data.Functor.Const
import Data.Proxy

class C2 a where
  m2 :: forall {k} {b :: k}. Const a b

newtype T a = MkT a
  deriving C2
``` 

It needs to bind `k`, `b` as `@{k} @{b}` in the derived method implementation but it can't do it because `@{k}`-binders are not part of the language today.

The proposal tries to address this problem by changing the `forall {k} {b :: k}.` syntax to stand for the new "quiet" specificity that can have `@{k}` binders. OK. But what about the new inferred binders that don't have the `@{{k}}` syntax? All we need to do to see that the issue hasn't gone anywhere is to update the example to use the new syntax of inferred binders

```haskell
import Data.Functor.Const
import Data.Proxy

class C2 a where
  m2 :: forall {{k}} {{b :: k}}. Const a b

newtype T a = MkT a
  deriving C2
``` 

and now we're back to where we started. There are no `@{{k}}` binders that the derived code could use.

simonpj commented:
> now we're back to where we started. There are no @{{k}} binders that the derived code could use.

No.  **The user can't write Inferred binders**.  And suppose in your `class C2` if you let GHC infer the binders, like this
```
class C3 a where
  m3 :: forall b. Const a b
```
Now indeed GHC will infer `m3 :: forall {{k}}. forall (b::k). Const @k a b`.  And that might limit what you can do with GND.    **But now you can fix it**, by explicitly quantifying the `k` as you do in your original class `C2`:
```
class C3 a where
  m3 :: forall {k}. forall (b::k). Const a b
```
(You could fix it today by writing
```
class C4 a where
  m4 :: forall k. forall (b::k). Const a b
```
But now type appplication of `m4` would need two type arguments.  The point of the proposal is to distinguish these two aspects.)

VitWW commented:
@simonpj You wrote wrong "Rendered" link (it points to just first commit) at the head comment.
This is right one: [Rendered](https://github.com/ghc-proposals/ghc-proposals/blob/wip/quiet-foralls/proposals/0000-quiet-foralls.rst)

> So if `f :: forall {k}. forall (a::k). Proxy a -> Int` we could call it thus:
> ```haskell
>f Proxy
> f @Int Proxy
> f @{Type} @Int Proxy
> f @{Type} Proxy
> ```
> That indeed seems plausible.

Yes, this is a key moment: `@{k}` (including `@{Type}`)  covers `TypeApplications`

> We attempted to explain this in the "TypeAbstractions" bullet of Section 1.1. Perhaps you can help us explain it better?

> `h :: forall {k1} {k2}. forall (a::k1) (b::k2). Proxy @k1 a -> Proxy @k2 b -> int`
> But does `k1` precede `k2`, or vice versa? The order would make a difference if we tried to define `h` thus: ...

First, Haskell has a rule: if we have a signature, what go first in `forall` (including default rules to infer all unwritten foralls), that go first in Type Abstraction (and Type Applications)

Second, we try to infer signature from Type Applications including `@{k}`

Third, we try to infer signature from Type Abstraction including `\ @{k} ->`

> Now the order of the k1/k2 foralls clearly matters. But GHC, not the user, has chosen that order, and there is no simple algorithm for specifying what order it should choose.

Fourth, if we have all signature and `@{k}` and `\ @{k} ->` we check if they match, otherwise it is an error.

VitWW commented:
> No. **The user can't write Inferred binders**. And suppose in your `class C2` if you let GHC infer the binders, like this
> 
> ```
> class C3 a where
>   m3 :: forall b. Const a b
> ```
> 
> Now indeed GHC will infer `m3 :: forall {{k}}. forall (b::k). Const @k a b`. And that might limit what you can do with GND. **But now you can fix it**, by explicitly quantifying the `k` as you do in your original class `C2`:
> 
> ```
> class C3 a where
>   m3 :: forall {k}. forall (b::k). Const a b
> ```

We could have a rule like in `ExplicitForall`: EITHER you omit to write "inferred" forall  (and GHC infer it instead of you) OR you write explicitly all "inferred foralls" in signature

simonpj commented:
> You wrote wrong "Rendered" link (it points to just first commit) at the first comment.

Thank you so much!  Fixed.

simonpj commented:
>   No. The user can't write Inferred binders.
> Aha! Now that's the actual payload of this proposal that fixes the issues. But is it spelled out in the proposal text?

Yes it was explicit in Section 2.  But I have now made it even more explicit.  Any better?

> Please do consider making @{} available at use sites.

I have done so.

You suggest Specified(default) and Specified(quiet).  Maybe... we are open to naming suggestions!

> Users shouldn't be able to write inferred binders. They're for the compiler to infer, period. But to address the original motivation of https://github.com/ghc-proposals/ghc-proposals/pull/99, you're introducing the "quiet" specificity.

Yes, precisely.  Is the text better now

Tritlo commented:
I had a hard time reading this proposal, but @int-index's clarification made it a lot clearer:

> * Required – user-accessible, must be bound and supplied visibly
> * Specified (default) – user-accessible, can be bound/supplied with `@`
> * Specified (quiet) – user-accessible, can be bound/supplied with `@{}`
> * Inferred – not user-accessible

The added braces for inferred: these are never written by the user, correct? Will they appear in error messages?

simonpj commented:
> but @int-index's clarification made it a lot clearer:

Great. I have adopted his phrasing. See if you like it better.

> The added braces for inferred: these are never written by the user, correct? 

Yes, that is stated even more explicitly.

> Will they appear in error messages?

That should be avoided, as it is today.  (Sometimes that might be impossible; but in any case it's not a new problem.)

> Edit: Maybe clarify that (if I understand correctly) that {{k}} is new, and {k} now means something else: the proposal makes it seem like we already had {{k}} and that we are adding {k}

I didn't understand that.  Maybe read the new Section 2 and see if it's any clearer?  If not can you suggest some extra wording?

Tritlo commented:
It's a lot clearer now, thanks!

goldfirere commented:
Summary:
 * Yes this is just a better version of #99.
 * Really Haskell should have named arguments, as discussed in #654.
 * Or Haskell should allow type variables to be given out of dependency order, as proposed in #131.
The current proposal is by far the easiest of these, but also the worst.

-----------------------------------------------------------------

I really like thinking of this proposal as a retraction of #99 and then a new plan to address the motivation of #99. At the time #99 was conceived, we had a mechanism -- inferred variables -- for making type variable abstractions that didn't change interfaces. So we just used the existing mechanism. But in the light of experience, we've learned that this doesn't quite work out (for the reasons described in this proposal). So we're now going to address #99 by making a new category of type variable binder, a quiet one.

One small thing is holding me back a bit, though. With quiet variables, we'll have two classes of specified variable. I can't see a reason why 2 is the right number though. One motivation is something like `typeRep :: forall {k} (t :: k). ...`, where we'll want to instantiate the `t` without bothering with the `k`. But maybe we have a function (I don't have an example) where we'll very often want to instantiate the third variable, sometimes the second and third, and rarely (but not never!) the first. Then someone will come along and make a proposal for "quieter" variables.

Thus perhaps a different way of seeing this -- and #99 -- is as a way of papering over Haskell's lack of named (type) arguments. That is, we have a natural way of being able to write a *prefix* of type arguments: include the arguments you want and then stop, letting GHC handle the rest. But we do not have a way of being able to write a *suffix* of type arguments, and thus we introduce quiet variables, making it so that users specify a prefix of *specified* variables (instead of all variables).

Interestingly, we don't play this game in terms. If I have `f :: A -> B -> C` and I want to apply `f` only to its argument of type `B`, I have to say `\a -> f a b` or `flip f b` or ``(`f` b)``. We've been OK with these for decades. Maybe we should just be OK with `typeRep @_ @Int`?

Yet another way to think of this all is that the problem really is that authors can't choose the order of type variables. That is, if my `f` frequently wanted to be applied only to its `B` argument, then I'd probably reorder to `f :: B -> A -> C`. But I can't do that with `typeRep`, because of its rigid requirement that type variables appear in dependency order. So an alternative to this proposal could be #131, which says right in its text that it obviates #99.

VitWW commented:
I still not convinced from Motivation the reason why split is needed from Inferred Forall into Quiet Forall and Inferred Forall.
But let's assume it is really needed.

My suggestion - have **same syntax** for both Quiet Forall and Inferred Forall, but to have a explicit rules where it is _forbidden_ to write explicitly `forall {k}`, `@{k}`, `\ @{k} ->` (for example for Left Hand variables of data/class declarations, etc)

int-index commented:
> Yes it was explicit in Section 2.

@simonpj You're right, it was there all along but I missed it. Anyway, it's now much clearer with the new edits.

Indeed, now I see the appeal of the proposed design.

> With quiet variables, we'll have two classes of specified variable. I can't see a reason why 2 is the right number though.

@goldfirere Yes, this is also bothering me. But I think it's fine to settle on `n=2` as the smallest number that allows us to avoid breakage. Ideally we'd go for `n=1` (i.e. to the pre-#99 design) plus named arguments, but then we'd end up breaking  user-written programs with `forall {k}.`, which is not nice. So realistically we should aim for `n=2` plus named arguments.

tomjaguarpaw commented on proposals/0000-quiet-foralls.rst:
```diff
@@ -0,0 +1,272 @@
+Here the type-lambda ``\ @b ->`` binds the type variable ``b`` in the body of the lambda,
+just as the term-lambda ``\(x::b) ->`` binds ``x`` in the body.  Proposal 425 makes a good
+case for the usefulness of this feature.  Note in particular tha the scoped
+variables arising from type signatures (using ``-XExetendedForAllScope``) are irrelevant here.
```
```suggestion
variables arising from type signatures (using ``-XExtendedForAllScope``) are irrelevant here.
```

tomjaguarpaw commented on proposals/0000-quiet-foralls.rst:
```diff
@@ -0,0 +1,272 @@
+
+Here the type-lambda ``\ @b ->`` binds the type variable ``b`` in the body of the lambda,
+just as the term-lambda ``\(x::b) ->`` binds ``x`` in the body.  Proposal 425 makes a good
+case for the usefulness of this feature.  Note in particular tha the scoped
```
```suggestion
case for the usefulness of this feature.  Note in particular that the scoped
```

tomjaguarpaw commented on proposals/0000-quiet-foralls.rst:
```diff
@@ -0,0 +1,272 @@
+=================================
+
+We propose to separate the two aspects of inferred foralls described above,
+by having *four* (instaed of three) forms for forall-binders:
```
```suggestion
by having *four* (instead of three) forms for forall-binders:
```

tomjaguarpaw commented:
I recognise the problem the proposal addresses. I have been stuck in situations where I wanted to apply an inferred type variable but couldn't. I can't actually remember where. I think it was something like

```.hs
type Some :: forall {k}. (k -> Type) -> Type
data Some f where
  MkSome :: SingI a => f a -> Some f
```

and I really wanted to write `Some @{k}` in some situation to get type class resolution to work out, or similar. Of course, if I'm using `Some f`, where the type of `f` is known, specifying `k` is redundant. But sometimes I want to use `Some` without an `f`!

I don't know if I can say for sure that the proposal provides the correct solution. Personally I'm prepared to put up with the pain until I feel like I better understand the problem and solution space.

phadej commented:
> Really Haskell should have named arguments

Sure, but I don't think this is *the* solution either.

Writing `Proxy @{a = Int}` is IMHO worse than `Proxy @_ @Int` (with `Proxy :: forall k -> k -> Type`), and clearly worse than `Proxy @Int` (with `Proxy :: forall {k} -> k -> Type`).

And I have to stress very much: if named arguments are enabled for all type-arguments, that makes **renaming any type variable a breaking change**. Given how inconsistent naming of variables are across the ecosystem it's a dangerous move to add named type variables. (For example [`Data.Map.Lazy`](https://hackage.haskell.org/package/containers-0.7/docs/Data-Map-Lazy.html) has most of the time `a` as value type, but sometimes `v` - it's fine as the names don't matter now).

There are already cases where people had to write tricky CPP to make `TypeApplications` - interface similar across different major GHC versions. I think the proposed change may 

---

@tomjaguarpaw 

> I don't know if I can say for sure that the proposal provides the correct solution.

I don't think there is "the" correct solution. Having "implicitness levels" is a way to allow skipping (type) arguments. Named arguments is another options (but see above).

The existence so called "inferred" type-variables, i.e. the fact that GHC does generalization in the (type or kind) inference is *the problem*, and one (not really acceptable) solution is just to remove that, i.e. never generalise.

konsumlamm commented on proposals/0000-quiet-foralls.rst:
```diff
@@ -0,0 +1,272 @@
+---------------------------------
+
+We do not propose to add a new extension flag; rather just to extend
+what ``-XTypeAbstractions`` does.  Old programs witll continue to
```
```suggestion
what ``-XTypeAbstractions`` does.  Old programs will continue to
```
